main.go

```diff
@@ -154,3 +189,247 @@ func gatekeeperNamespace(platformInfo platform.PlatformInfo) (string, error) {
+
+		return false
+	}
+
+	for _, condition := range conditions {
+		parsedCondition := condition.(map[string]interface{})
+
+		status, ok := parsedCondition["status"].(string)
+		if !ok {
+			setupLog.V(1).Info("Cannot parse ConstraintPodStatus conditions status")
+
+			return false
+		}
+
+		conditionType, ok := parsedCondition["type"].(string)
+		if !ok {
+			setupLog.V(1).Info("Cannot parse ConstraintPodStatus conditions type")
+
+			return false
+		}
+
+		if conditionType == "NamesAccepted" && status == "True" {
+			setupLog.V(1).Info("ConstraintPodStatus crd is ready")
+
+			return true
+		}
+	}
+
+	setupLog.V(1).Info("ConstraintPodStatus crd is not ready yet")
+
+	return false
+}
+
+// Check gatekeeper auditFromCache=Automatic
+// Check constraintpodstatuses.status.gatekeeper.sh crd status has "NamesAccepted" condition that is true.
+func checkCPScontrollerAvailable(mainCtx context.Context, result watch.Event, dynamicClient *dynamic.DynamicClient) bool {
+	gatekeeper, ok := runtime.DefaultUnstructuredConverter.ToUnstructured(result.Object)
+	if ok != nil {
+		return false
+	}
+
+	audit, _, err := unstructured.NestedMap(gatekeeper, "spec", "audit")
+	if err != nil {
+		setupLog.V(1).Info("audit is not set. ConstraintStatus controller cannot be created")
+		return false
+	}
+
+	auditFromCache := audit["auditFromCache"]
+	if auditFromCache != string(operatorv1alpha1.AuditFromCacheAutomatic) {
+		setupLog.V(1).Info("auditFromCache is not Automatic. ConstraintStatus controller cannot be created")
+
+		return false
+	}
+
+	i := 0
+	for i < 10 {
+		isCPSCrdExist := checkCPSCrdAvailable(mainCtx, dynamicClient)
+		if isCPSCrdExist {
+			break
+		}
+		// It has been waited for around 30 seconds but still not ready
+		if i == 9 && !isCPSCrdExist {
+			setupLog.V(1).Info("ConstraintPodStatus crd is still not ready. return checkCPScontrollerAvailable to false")
+			return false
+		}
+
+		time.Sleep(time.Second * 3)
+
+		i++
+	}
+
+	return true
+}
+
+// Watch the gatekeeper resource in this fuction.
+// This function adds constraintpodstatus controller when Adding(Creation) and Modified events happens
+// When deleting event comes, close the controller using context
+func watchGatekeeperInstall(mainCtx context.Context, dynamicClient *dynamic.DynamicClient,
+	cfg *rest.Config, enableLeaderElection bool, namespace string,
+) {
+	ctx, ctxCancel := context.WithCancel(mainCtx)
+	isInstalled := false
+
+	watcher, err := newWatcher(mainCtx, dynamicClient)
+	if err != nil {
+		setupLog.Error(err, "Error to setup watcher")
+		ctxCancel()
+		return
+	}
+
+	for {
+		select {
+		case result, ok := <-watcher.ResultChan():
+			// the channel got closed, so we need to restart
+			if !ok {
+				watcher, err = newWatcher(mainCtx, dynamicClient)
+				if err != nil {
+					setupLog.Error(err, "Error to setup watcher")
+					ctxCancel()
+					return
+				}
+
+				continue
+			}
+
+			switch result.Type { //nolint:exhaustive
+			case watch.Modified, watch.Added:
```

Comment:
What about when the user turns off automatic? What happens to the SyncOnly values in the `Config` object? It seems to me that if it's set to `Enabled` from `Automatic` that it should be wiped out.

This would be easier if you follow this comment:
https://github.com/stolostron/gatekeeper-operator/pull/18/files?diff=split&w=0#r1393341166

Reply:
In my opinion, wiping out existing `syncOnly` is dangerous. Maybe they want existing syncOnly but they want to add something. Also, it means a sudden change user's env. If users don't want `syncOnly` they can go to config and remove them